File: scripts/native-runtime-resolution-regression.swift
import Foundation

enum NativeRuntimeResolutionRegressionFailure: Error {
  case assertion(String)
}

func requireRuntimeResolution(_ condition: @autoclosure () -> Bool, _ message: String) throws {
  if !condition() { throw NativeRuntimeResolutionRegressionFailure.assertion(message) }
}

@main
struct NativeRuntimeResolutionRegression {
  static func main() throws {
    let home = URL(fileURLWithPath: "/Users/tester")
    let executablePaths = Set([
      "/Users/tester/.local/bin/codex",
      "/opt/homebrew/bin/opl",
      "/custom/codex"
    ])
    let executable: (String) -> Bool = { executablePaths.contains($0) }

    let finderCodex = resolveExternalExecutable(
      name: "codex",
      explicitEnvironmentKeys: ["OPL_CODEX_BIN", "CODEX_CLI_PATH", "CODEX_BIN"],
      environment: ["PATH": "/usr/bin:/bin"],
      homeDirectory: home,
      isExecutable: executable
    )
    try requireRuntimeResolution(finderCodex == "/Users/tester/.local/bin/codex", "Finder launch must resolve ~/.local/bin/codex")

    let explicitCodex = resolveExternalExecutable(
      name: "codex",
      explicitEnvironmentKeys: ["OPL_CODEX_BIN", "CODEX_CLI_PATH", "CODEX_BIN"],
      environment: ["OPL_CODEX_BIN": "/custom/codex", "PATH": "/usr/bin:/bin"],
      homeDirectory: home,
      isExecutable: executable
    )
    try requireRuntimeResolution(explicitCodex == "/custom/codex", "OPL_CODEX_BIN must remain the highest priority")

    let opl = resolveExternalExecutable(
      name: "opl",
      explicitEnvironmentKeys: ["OPL_APP_OPL_BIN"],
      environment: ["PATH": "/usr/bin:/bin"],
      homeDirectory: home,
      isExecutable: executable
    )
    try requireRuntimeResolution(opl == "/opt/homebrew/bin/opl", "Finder launch must resolve Homebrew OPL")

    let candidates = externalExecutableCandidates(
      name: "codex",
      explicitEnvironmentKeys: ["OPL_CODEX_BIN", "CODEX_CLI_PATH", "CODEX_BIN"],
      environment: ["PATH": "/usr/bin:/bin"],
      homeDirectory: home
    )
    try requireRuntimeResolution(candidates.allSatisfy { !$0.localizedCaseInsensitiveContains("aioncore") }, "AionCore must never be a Native executable candidate")

    print("{\"status\":\"native_runtime_resolution_regression_passed\"}")
  }
}

File: scripts/native-runtime-resolution-regression.mjs
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

const root = path.resolve(new URL("..", import.meta.url).pathname);
const nativePath = path.join(root, "scripts/opl-studio-app.swift");
const source = fs.readFileSync(nativePath, "utf8");
const bootstrapIndex = source.indexOf("\nlet app = NSApplication.shared");
assert(bootstrapIndex >= 0, "native bootstrap marker");

for (const marker of [
  "resolveExternalExecutable(",
  'explicitEnvironmentKeys: ["OPL_CODEX_BIN", "CODEX_CLI_PATH", "CODEX_BIN"]',
  'homeDirectory.appendingPathComponent(".local/bin/codex")',
  '"/opt/homebrew/bin/\\(name)"',
  "private let initializationLock = NSLock()",
  "initializationLock.lock()",
  "private let appServer: CodexAppServerClient"
]) assert(source.includes(marker), `missing native runtime resolution marker ${marker}`);

assert.equal(source.toLowerCase().includes("bundled-aioncore"), false, "Native must not resolve a bundled AionCore runtime");

const tempRoot = fs.mkdtempSync(path.join(os.tmpdir(), "opl-native-runtime-resolution-"));
try {
  const testSource = path.join(tempRoot, "OplStudioTestSource.swift");
  const executable = path.join(tempRoot, "native-runtime-resolution-regression");
  fs.writeFileSync(testSource, source.slice(0, bootstrapIndex));
  const compile = spawnSync("swiftc", [
    testSource,
    path.join(root, "scripts/native-runtime-resolution-regression.swift"),
    "-framework",
    "Cocoa",
    "-framework",
    "WebKit",
    "-o",
    executable
  ], { encoding: "utf8", cwd: root });
  assert.equal(compile.status, 0, `native runtime resolution compile failed\n${compile.stdout}\n${compile.stderr}`);
  const run = spawnSync(executable, [], { encoding: "utf8", cwd: root });
  assert.equal(run.status, 0, `native runtime resolution regression failed\n${run.stdout}\n${run.stderr}`);
  process.stdout.write(run.stdout);
} finally {
  fs.rmSync(tempRoot, { recursive: true, force: true });
}
